File: examples/client4.c
/*
 * Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
 *                         University Research and Technology
 *                         Corporation.  All rights reserved.
 * Copyright (c) 2004-2011 The University of Tennessee and The University
 *                         of Tennessee Research Foundation.  All rights
 *                         reserved.
 * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
 *                         University of Stuttgart.  All rights reserved.
 * Copyright (c) 2004-2005 The Regents of the University of California.
 *                         All rights reserved.
 * Copyright (c) 2006-2013 Los Alamos National Security, LLC.
 *                         All rights reserved.
 * Copyright (c) 2009-2012 Cisco Systems, Inc.  All rights reserved.
 * Copyright (c) 2011      Oak Ridge National Labs.  All rights reserved.
 * Copyright (c) 2013-2020 Intel, Inc.  All rights reserved.
 * Copyright (c) 2015      Mellanox Technologies, Inc.  All rights reserved.
 * Copyright (c) 2019      IBM Corporation.  All rights reserved.
 * Copyright (c) 2021-2025 Nanook Consulting  All rights reserved.
 * Copyright (c) 2022      ParTec AG.  All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 *
 */

#define _GNU_SOURCE
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

#include "examples.h"
#include <pmix.h>

int main(int argc, char **argv)
{
    pmix_proc_t myproc;
    pmix_status_t rc;
    pmix_value_t *val = NULL;
    pmix_proc_t proc;

    EXAMPLES_HIDE_UNUSED_PARAMS(argc, argv);

    if (PMIX_SUCCESS != (rc = PMIx_Init(&myproc, NULL, 0))) {
        if (PMIX_ERR_UNREACH == rc) {
            fprintf(stderr, "%s: Cannot operate as singleton\n",
                    PMIx_Proc_string(&myproc));
        } else {
            fprintf(stderr, "%s: PMIx_Init failed: %s\n",
                    PMIx_Proc_string(&myproc),
                    PMIx_Error_string(rc));
        }
        exit(1);
    }
    fprintf(stderr, "%s: Running\n", PMIx_Proc_string(&myproc));

    // get our rank
    PMIX_LOAD_PROCID(&proc, myproc.nspace, PMIX_RANK_INVALID);
    rc = PMIx_Get(&proc, PMIX_RANK, NULL, 0, &val);
    if (PMIX_SUCCESS != rc) {
        fprintf(stderr, "%s: Get rank failed - %s\n", PMIx_Proc_string(&myproc), PMIx_Error_string(rc));
        fflush(stderr);
        goto done;
    }
    if (val->data.rank != myproc.rank) {
        fprintf(stderr, "%s: Get rank returned wrong rank - %u instead of %u\n",
                PMIx_Proc_string(&myproc), val->data.rank, myproc.rank);
        fflush(stderr);
        goto done;
    }
    fprintf(stderr, "%s: Rank return correct\n", PMIx_Proc_string(&myproc));
    PMIX_VALUE_RELEASE(val);

    // get our node ID
    rc = PMIx_Get(&myproc, PMIX_NODEID, NULL, 0, &val);
    if (PMIX_SUCCESS != rc) {
        fprintf(stderr, "%s: Get my nodeID failed - %s\n", PMIx_Proc_string(&myproc), PMIx_Error_string(rc));
        fflush(stderr);
        goto done;
    }
    // rank=0 is on node 0, rank=1 on node 1
    if (val->data.uint32 != myproc.rank) {
        fprintf(stderr, "%s: Get my nodeID returned wrong value - %u instead of %u\n",
                PMIx_Proc_string(&myproc), val->data.uint32, myproc.rank);
        fflush(stderr);
        goto done;
    }
    fprintf(stderr, "%s: NodeID return correct\n", PMIx_Proc_string(&myproc));
    PMIX_VALUE_RELEASE(val);

    // get our node ID with rank=WILDCARD - should fail!
    PMIX_LOAD_PROCID(&proc, myproc.nspace, PMIX_RANK_WILDCARD);
    rc = PMIx_Get(&proc, PMIX_NODEID, NULL, 0, &val);
    if (PMIX_SUCCESS == rc) {
        fprintf(stderr, "%s: Get my nodeID with WILDCARD rank incorrectly succeeded\n",
                PMIx_Proc_string(&myproc));
        fflush(stderr);
        goto done;
    }
    fprintf(stderr, "%s: NodeID with WILDCARD rank correctly failed - %s\n",
            PMIx_Proc_string(&myproc), PMIx_Error_string(rc));
    PMIX_VALUE_RELEASE(val);

    // get our peer's nodeID
    if (0 == myproc.rank) {
        proc.rank = 1;
    } else {
        proc.rank = 0;
    }
    rc = PMIx_Get(&proc, PMIX_NODEID, NULL, 0, &val);
    if (PMIX_SUCCESS != rc) {
        fprintf(stderr, "%s: Get peer's nodeID failed - %s\n", PMIx_Proc_string(&myproc), PMIx_Error_string(rc));
        fflush(stderr);
        goto done;
    }
    // rank=0 is on node 0, rank=1 on node 1
    if (val->data.uint32 != proc.rank) {
        fprintf(stderr, "%s: Get peer's nodeID returned wrong value - %u instead of %u\n",
                PMIx_Proc_string(&myproc), val->data.uint32, proc.rank);
        fflush(stderr);
        goto done;
    }
    fprintf(stderr, "%s: Peer's NodeID return correct\n", PMIx_Proc_string(&myproc));
    PMIX_VALUE_RELEASE(val);

    // get our appnum
    rc = PMIx_Get(&myproc, PMIX_APPNUM, NULL, 0, &val);
    if (PMIX_SUCCESS != rc) {
        fprintf(stderr, "%s: Get my appnum failed - %s\n", PMIx_Proc_string(&myproc), PMIx_Error_string(rc));
        fflush(stderr);
        goto done;
    }
    // rank=0 is on node 0, rank=1 on node 1
    if (val->data.uint32 != myproc.rank) {
        fprintf(stderr, "%s: Get my appnum returned wrong value - %u instead of %u\n",
                PMIx_Proc_string(&myproc), val->data.uint32, myproc.rank);
        fflush(stderr);
        goto done;
    }
    fprintf(stderr, "%s: Appnum return correct\n", PMIx_Proc_string(&myproc));
    PMIX_VALUE_RELEASE(val);

    // get appnum with WILDCARD - should fail!
    PMIX_LOAD_PROCID(&proc, myproc.nspace, PMIX_RANK_WILDCARD);
    rc = PMIx_Get(&proc, PMIX_APPNUM, NULL, 0, &val);
    if (PMIX_SUCCESS == rc) {
        fprintf(stderr, "%s: Get appnum with WILDCARD incorrectly succeeded - returned %u\n",
                PMIx_Proc_string(&myproc), val->data.uint32);
        fflush(stderr);
        goto done;
    }
    fprintf(stderr, "%s: Appnum with WILDCARD rank correctly failed - %s\n",
            PMIx_Proc_string(&myproc), PMIx_Error_string(rc));
    PMIX_VALUE_RELEASE(val);

    // get peer's appnum
    if (0 == myproc.rank) {
        proc.rank = 1;
    } else {
        proc.rank = 0;
    }
    rc = PMIx_Get(&proc, PMIX_APPNUM, NULL, 0, &val);
    if (PMIX_SUCCESS != rc) {
        fprintf(stderr, "%s: Get peer's appnum failed - %s\n", PMIx_Proc_string(&myproc), PMIx_Error_string(rc));
        fflush(stderr);
        goto done;
    }
    // rank=0 is in appnum 0, rank=1 is in appnum 1
    if (val->data.uint32 != proc.rank) {
        fprintf(stderr, "%s: Get peer's appnum returned wrong value - %u instead of %u\n",
                PMIx_Proc_string(&myproc), val->data.uint32, proc.rank);
        fflush(stderr);
        goto done;
    }
    fprintf(stderr, "%s: Peer's Appnum return correct\n", PMIx_Proc_string(&myproc));
    PMIX_VALUE_RELEASE(val);

done:
    /* finalize us */
    rc = PMIx_Finalize(NULL, 0);
    fflush(stderr);
    return (0);
}

File: examples/python/client4.py
#!/usr/bin/env python3
#
# Copyright (c) 2026      Nanook Consulting  All rights reserved.
# $COPYRIGHT$
#
# Additional copyrights may follow
#
# $HEADER$
#
# Python port of examples/client4.c
#
# Check the rank-qualified job info: rank, nodeID and appnum must come
# back correctly for ourselves and for a peer, and must be refused when
# asked for with the wildcard rank (they are per-rank values, so there is
# no job-level answer).
#
# Expects two procs, one per node, each in its own app.

import sys

from examples import *


def main():
    client = PMIxClient()

    rc, myproc = client.init(None)
    if PMIX_SUCCESS != rc:
        if PMIX_ERR_UNREACH == rc:
            eprint("Cannot operate as singleton")
        else:
            eprint("PMIx_Init failed:", client.error_string(rc))
        sys.exit(1)
    # our identity is fixed, so cache its string form and reuse it. Unlike
    # the C version there is nothing to free - proc_string() hands back an
    # ordinary Python string
    _, me = client.proc_string(myproc)
    eprint("%s: Running" % me)

    # get our rank
    proc = {'nspace': myproc['nspace'], 'rank': PMIX_RANK_INVALID}
    rc, val = client.get(proc, PMIX_RANK, None)
    if PMIX_SUCCESS != rc:
        eprint("%s: Get rank failed - %s" % (me, client.error_string(rc)))
        return done(client)
    if val['value'] != myproc['rank']:
        eprint("%s: Get rank returned wrong rank - %u instead of %u"
               % (me, val['value'], myproc['rank']))
        return done(client)
    eprint("%s: Rank return correct" % me)

    # get our node ID
    rc, val = client.get(myproc, PMIX_NODEID, None)
    if PMIX_SUCCESS != rc:
        eprint("%s: Get my nodeID failed - %s" % (me, client.error_string(rc)))
        return done(client)
    # rank=0 is on node 0, rank=1 on node 1
    if val['value'] != myproc['rank']:
        eprint("%s: Get my nodeID returned wrong value - %u instead of %u"
               % (me, val['value'], myproc['rank']))
        return done(client)
    eprint("%s: NodeID return correct" % me)

    # get our node ID with rank=WILDCARD - should fail!
    proc = {'nspace': myproc['nspace'], 'rank': PMIX_RANK_WILDCARD}
    rc, val = client.get(proc, PMIX_NODEID, None)
    if PMIX_SUCCESS == rc:
        eprint("%s: Get my nodeID with WILDCARD rank incorrectly succeeded"
               % me)
        return done(client)
    eprint("%s: NodeID with WILDCARD rank correctly failed - %s"
           % (me, client.error_string(rc)))

    # get our peer's nodeID
    proc['rank'] = 1 if 0 == myproc['rank'] else 0
    rc, val = client.get(proc, PMIX_NODEID, None)
    if PMIX_SUCCESS != rc:
        eprint("%s: Get peer's nodeID failed - %s"
               % (me, client.error_string(rc)))
        return done(client)
    # rank=0 is on node 0, rank=1 on node 1
    if val['value'] != proc['rank']:
        eprint("%s: Get peer's nodeID returned wrong value - %u instead of %u"
               % (me, val['value'], proc['rank']))
        return done(client)
    eprint("%s: Peer's NodeID return correct" % me)

    # get our appnum
    rc, val = client.get(myproc, PMIX_APPNUM, None)
    if PMIX_SUCCESS != rc:
        eprint("%s: Get my appnum failed - %s" % (me, client.error_string(rc)))
        return done(client)
    # rank=0 is in appnum 0, rank=1 is in appnum 1
    if val['value'] != myproc['rank']:
        eprint("%s: Get my appnum returned wrong value - %u instead of %u"
               % (me, val['value'], myproc['rank']))
        return done(client)
    eprint("%s: Appnum return correct" % me)

    # get appnum with WILDCARD - should fail!
    proc = {'nspace': myproc['nspace'], 'rank': PMIX_RANK_WILDCARD}
    rc, val = client.get(proc, PMIX_APPNUM, None)
    if PMIX_SUCCESS == rc:
        eprint("%s: Get appnum with WILDCARD incorrectly succeeded - "
               "returned %u" % (me, val['value']))
        return done(client)
    eprint("%s: Appnum with WILDCARD rank correctly failed - %s"
           % (me, client.error_string(rc)))

    # get peer's appnum
    proc['rank'] = 1 if 0 == myproc['rank'] else 0
    rc, val = client.get(proc, PMIX_APPNUM, None)
    if PMIX_SUCCESS != rc:
        eprint("%s: Get peer's appnum failed - %s"
               % (me, client.error_string(rc)))
        return done(client)
    # rank=0 is in appnum 0, rank=1 is in appnum 1
    if val['value'] != proc['rank']:
        eprint("%s: Get peer's appnum returned wrong value - %u instead of %u"
               % (me, val['value'], proc['rank']))
        return done(client)
    eprint("%s: Peer's Appnum return correct" % me)

    return done(client)


def done(client):
    # finalize us
    client.finalize(None)
    return 0


if __name__ == '__main__':
    sys.exit(main())
